### PR TITLE
granian: init at 1.5.2

### LINKED ISSUE #337480: Package request: granian
**Project description**

Granian is a Rust HTTP server for Python ASGI/WSGI/RSGI applications.
It fulfills a similar purpose as daphne, gunicorn, uvicorn or uwsgi, but [seems to perform better](https://github.com/emmett-framework/granian/blob/master/benchmarks/vs.md).

**Metadata**

* homepage URL: https://github.com/emmett-framework/granian
* source URL: https://github.com/emmett-framework/granian
* license: bsd3
* platforms: unix

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by lucastso10**
## Description of changes

Added the [Granian](https://github.com/emmett-framework/granian) app

fixes #337480

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/gr/granian/package.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+}:
+
+python3.pkgs.buildPythonApplication rec {
+  pname = "granian";
+  version = "1.5.2";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "emmett-framework";
+    repo = pname;
```
Comment (eclairevoyant) [inline pkgs/by-name/gr/granian/package.nix:16]:
```suggestion
    repo = "granian";
```

`pname` in URLs is discouraged, and it's overused elsewhere as well

File: pkgs/by-name/gr/granian/package.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    maturinBuildHook
+  ];
+
+  dependencies = [
+    python312Packages.uvloop
+    python312Packages.click
+  ];
+
+  meta = {
+    description = "A Rust HTTP server for Python ASGI/WSGI/RSGI applications.";
```
Comment (eclairevoyant) [inline pkgs/by-name/gr/granian/package.nix:39]:
```suggestion
    description = "Rust HTTP server for Python ASGI/WSGI/RSGI applications";
```

no leading article, no trailing period, per guidelines

File: pkgs/by-name/gr/granian/package.nix [hunk 3]
```diff
@@ -0,0 +1,46 @@
+{
+  lib,
+  python3,
+  fetchFromGitHub,
+  rustPlatform,
+  python312Packages,
```
Comment (eclairevoyant) [inline pkgs/by-name/gr/granian/package.nix:6]:
```suggestion
  python3Packages,
```

File: pkgs/by-name/gr/granian/package.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
+{
+  lib,
+  python3,
```
Comment (eclairevoyant) [inline pkgs/by-name/gr/granian/package.nix:3]:
```suggestion
```

File: pkgs/by-name/gr/granian/package.nix [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  fetchFromGitHub,
+  rustPlatform,
+  python312Packages,
+}:
+
+python3.pkgs.buildPythonApplication rec {
```
Comment (eclairevoyant) [inline pkgs/by-name/gr/granian/package.nix:9]:
```suggestion
python3Packages.buildPythonApplication rec {
```

File: pkgs/by-name/gr/granian/package.nix [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  };
+
+  nativeBuildInputs = with rustPlatform; [
+    cargoSetupHook
+    maturinBuildHook
+  ];
+
+  dependencies = [
+    python312Packages.uvloop
+    python312Packages.click
```
Comment (eclairevoyant) [inline pkgs/by-name/gr/granian/package.nix:35]:
```suggestion
    python3Packages.uvloop
    python3Packages.click
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

**Comment by lucastso10** [comment]
@eclairevoyant Thank you for the review, I've already applied the requested changes!

File: pkgs/by-name/gr/granian/package.nix [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    maturinBuildHook
+  ];
+
+  dependencies = [
+    python3Packages.uvloop
+    python3Packages.click
+  ];
+
+  meta = {
+    description = "Rust HTTP server for Python ASGI/WSGI/RSGI applications.";
```
Comment (eclairevoyant) [inline pkgs/by-name/gr/granian/package.nix:38]:
```suggestion
    description = "Rust HTTP server for Python ASGI/WSGI/RSGI applications";
```

Comment (lucastso10) [reply pkgs/by-name/gr/granian/package.nix:38]: oops, sorry

**Review by eclairevoyant — APPROVED** [review]

**Comment by Aleksanaa** [comment]
@NixOS/darwin-maintainers please try to fix (now we need `libiconv` in `buildInputs`, but maybe others)

**Comment by prusnak** [comment]
I just confirmed that adding `libiconv` to `buildInputs` fixes the issue on Darwin:

``` patch
--- package.nix
+++ package.nix
@@ -1,7 +1,9 @@
 {
   lib,
+  stdenv,
   fetchFromGitHub,
   rustPlatform,
+  libiconv,
   python3Packages,
 }:
 
@@ -29,6 +31,10 @@ python3Packages.buildPythonApplication rec {
     maturinBuildHook
   ];
 
+  buildInputs = lib.optionals stdenv.isDarwin [
+    libiconv
+  ];
+
   dependencies = [
     python3Packages.uvloop
     python3Packages.click
```

**Comment by emilazy** [comment]
No need for the conditional; `libiconv` is `null` on platforms that don’t need it.

**Comment by lucastso10** [comment]
Sorry for the delay, and thank you for the work everybody!

**Comment by lucastso10** [comment]
I've added libconv whitout the conditional as @emilazy suggested!

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 337746` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>granian</li>
    <li>granian.dist</li>
  </ul>
</details>

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 337746` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>granian</li>
    <li>granian.dist</li>
  </ul>
</details>

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**